### PR TITLE
GH-36284: [Python][Parquet] Support write page index in Python API

### LINKED ISSUE #36284: [Python][Parquet] Support write page index in Parquet
### Describe the enhancement requested

Currently, Parquet support write page-index, now we want parquet Python api could write page index.

### Component(s)

Parquet, Python

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Support `write_page_index` in Parquet Python API

### What changes are included in this PR?

support `write_page_index` in properties

### Are these changes tested?

Currently not

### Are there any user-facing changes?

User can generate page index here.

* Closes: #36284

### CODE REVIEW

**Comment by mapleFU** [comment]
@jorisvandenbossche Mind take a look? I'm not so familiar with Python part, so maybe make something wrong

**Comment by mapleFU** [comment]
@pitrou @westonpace Would you mind take a look? This patch support Python to write `page_index`.

File: python/pyarrow/parquet/core.py
Comment (AlenkaF) [inline python/pyarrow/parquet/core.py:873]:
```suggestion
    Parquet format supports page index that makes filtering when
    reading more efficient. This option enables writing the page
    index to the Parquet file.
```

**Review by AlenkaF — COMMENTED** [review]
I only have a minor suggestion about the `write_page_index` docstrings.

**Review by AlenkaF — APPROVED** [review]
Thanks!

**Comment by mapleFU** [comment]
Can this patch be merged? Or should I wait for other committers review?

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by mapleFU** [comment]
* test-conda-python-3.10-spark-master
* test-cuda-python
* test-conda-python-3.8-spark-v3.1.2
* test-conda-python-3.10-spark-master

These cases failed, how can I try to fix them?

**Comment by pitrou** [comment]
@mapleFU Those are unrelated to this PR. Can you try to rebase?

**Comment by AlenkaF** [comment]
* test-conda-python-3.10-spark-master
* test-conda-python-3.8-spark-v3.1.2
* test-conda-python-3.9-spark-v3.2.0

Spark failures are known and have an issue opened.

* test-conda-python-3.11-hypothesis

Hypothesis failure is a new one but I do not see how it could be related to this PR.

* test-cuda-python

I have seen nightlies fail with this error today already, so this is not related to the PR either.

**Comment by jorisvandenbossche** [comment]
> Hypothesis failure is a new one but I do not see how it could be related to this PR.

Hmm, that seems very similar to the one that I fixed last week (https://github.com/apache/arrow/issues/36349, but now with another unknown timezone). In any case, you can ignore it here.

File: python/pyarrow/parquet/core.py
Comment (jorisvandenbossche) [inline python/pyarrow/parquet/core.py:870]: Side question: should we consider making this turned on by default at some point?

**Review by jorisvandenbossche — APPROVED** [review]

File: python/pyarrow/_parquet.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_parquet.pyx:498]: A actual question for this PR: we already do have a `has_index_page` attribute. How does that exactly differ?

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

File: python/pyarrow/_parquet.pyx
Comment (pitrou) [reply python/pyarrow/_parquet.pyx:498]: Also it would be nice to implement `has_index_page` and `index_page_offset`...

Comment (mapleFU) [reply python/pyarrow/_parquet.pyx:498]: To be short, `IndexPage` is not `PageIndex`.

1. https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L536 Currently, index page turns to be something that user defined.
2. PageIndex is a zonemap for pages

So they are different things

File: python/pyarrow/parquet/core.py
Comment (mapleFU) [reply python/pyarrow/parquet/core.py:870]: Currently not, I found it's hard to implement page index pruning in current implementions. If we implements it, maybe we can change it to default.

Comment (pitrou) [reply python/pyarrow/parquet/core.py:870]: Even if it's  not used already, it would probably be beneficial to write files with the index enabled, for future use.
Is there a performance issue with enabling it?

File: python/pyarrow/_parquet.pyx
Comment (pitrou) [reply python/pyarrow/_parquet.pyx:498]: Ahah, I forgot about that.

File: python/pyarrow/parquet/core.py
Comment (mapleFU) [reply python/pyarrow/parquet/core.py:870]: I guess most time there is no performance issue. But when user has extremly long string, we might write to much data.

Comment (pitrou) [reply python/pyarrow/parquet/core.py:870]: We are allowed to trim the min/max values, right?

**Comment by mapleFU** [comment]
Still these failed, lol

File: python/pyarrow/tests/parquet/test_metadata.py
Comment (pitrou) [inline python/pyarrow/tests/parquet/test_metadata.py:371]: Ok, but can we also have a test where these properties are false?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: python/pyarrow/parquet/core.py
Comment (jorisvandenbossche) [reply python/pyarrow/parquet/core.py:870]: So if I understand correctly, we are currently not yet using the PageIndex when reading files (through the python APIs) for pruning pages when given a filter? 

Should we mention that in the docstring to note that you can already write a PageIndex, but it will not yet be used when reading using pyarrow?

File: python/pyarrow/_parquet.pyx
Comment (jorisvandenbossche) [reply python/pyarrow/_parquet.pyx:1615]: I think this pattern comes from the code above, where such isinstance check is also used. But that's for cases where the keyword could either be True/False or a dictionary to enable the option per column. 
Here it's just a single True/False, so indeed the check is not needed. 

I think we can simply leave out the `if isinstance` and just have the ``if write_page_index: ... else: ...``

Comment (mapleFU) [reply python/pyarrow/_parquet.pyx:1615]: long time not written Python so just reference the code style above. I'll change it.

File: python/pyarrow/parquet/core.py
Comment (mapleFU) [reply python/pyarrow/parquet/core.py:870]: @jorisvandenbossche I've done that. By the way, we cannot filter using pyarrow, but `parquet-rs` and `parquet-mr` can optimize by it.

File: python/pyarrow/tests/parquet/test_metadata.py
Comment (mapleFU) [reply python/pyarrow/tests/parquet/test_metadata.py:371]: done

File: python/pyarrow/_parquet.pyx
Comment (pitrou) [reply python/pyarrow/_parquet.pyx:1615]: You didn't change anything, did you?

Comment (mapleFU) [reply python/pyarrow/_parquet.pyx]: Sorry...maybe I forgot to pop the stash

**Comment by mapleFU** [comment]
@pitrou @jorisvandenbossche I've tried to fix the comment here. Would you mind take a look?

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Pull request merged**
